File: Games/Uzta/src/main/kotlin/jokatu/game/games/uzta/event/ScoreUpdatedEventHandler.java
package jokatu.game.games.uzta.event;

import jokatu.game.event.AbstractEventHandler;
import jokatu.game.games.uzta.game.Uzta;
import jokatu.game.games.uzta.graph.LineSegment;
import jokatu.game.games.uzta.graph.UztaGraph;
import jokatu.game.games.uzta.player.UztaPlayer;
import org.jetbrains.annotations.NotNull;
import org.springframework.stereotype.Component;

import java.text.MessageFormat;

import static java.util.function.Predicate.isEqual;

/**
 * When the {@link GraphUpdatedEvent graph is updated}, update each player's score.
 */
@Component
public class ScoreUpdatedEventHandler extends AbstractEventHandler<Uzta, GraphUpdatedEvent> {
	@NotNull
	@Override
	protected Class<GraphUpdatedEvent> getEventClass() {
		return GraphUpdatedEvent.class;
	}

	@NotNull
	@Override
	protected Class<Uzta> getGameClass() {
		return Uzta.class;
	}

	@Override
	protected void handleCastGameAndEvent(
			@NotNull Uzta game, @NotNull GraphUpdatedEvent event
	) {
		game.getPlayers().forEach(
				player -> sender.send(
						MessageFormat.format(
								"/topic/score.game.{0}.{1}",
								game.getIdentifier(),
								player.getName()
						),
						calculateScore(player, game.getGraph())
				)
		);
	}

	private long calculateScore(@NotNull UztaPlayer player, @NotNull UztaGraph graph) {
		return graph.getEdges().stream()
				.map(LineSegment::getOwner)
				.filter(isEqual(player))
				.count();
	}
}

File: Games/Uzta/src/main/kotlin/jokatu/game/games/uzta/event/ScoreUpdatedEventHandler.kt
package jokatu.game.games.uzta.event

import jokatu.components.stomp.PlayerScore
import jokatu.game.event.AbstractEventHandler
import jokatu.game.games.uzta.game.Uzta
import jokatu.game.games.uzta.graph.UztaGraph
import jokatu.game.games.uzta.player.UztaPlayer
import ophelia.kotlin.number.AnnotatedInt
import ophelia.kotlin.number.AnnotatedSum
import ophelia.kotlin.number.IntFrom
import org.springframework.stereotype.Component
import java.util.function.Predicate.isEqual
import java.util.stream.Collectors.toSet

/**
 * When the [graph is updated][GraphUpdatedEvent], update each player's score.
 */
@Component
class ScoreUpdatedEventHandler : AbstractEventHandler<Uzta, GraphUpdatedEvent>(Uzta::class, GraphUpdatedEvent::class) {
	override fun handleCastGameAndEvent(game: Uzta, event: GraphUpdatedEvent) {
		game.getPlayers().forEach { player ->
			sender.send(PlayerScore(game, player), calculateScore(player, game.finalisedGraph))
		}
	}

	private fun calculateScore(player: UztaPlayer, graph: UztaGraph): AnnotatedInt {
		val ownedEdges = graph.edges.stream()
				.map { it.owner }
				.filter(isEqual(player))
				.collect(toSet())

		val ownedEdgesNumber = IntFrom(ownedEdges.size, "from owned edges")

		return AnnotatedSum(
				ownedEdgesNumber
		)
	}
}